File: linked_list/has_cycle.c
#include <stdbool.h>
#include <stdlib.h>

/**
 * Structure for the Singly Linked List
*/
typedef struct SinglyLinkedListNode {
    int data;
    SinglyLinkedListNode* next;
}SinglyLinkedListNode;

/**
 * has_cycle - finds if a linked list has a cycle using
 * Floyd's Cycle Detection Algorithm
 * @head: head of the linked list
 *
 * Return: If the linked list has a cycle, it returns true.
 * Otherwise, it returns false.
*/
bool has_cycle(SinglyLinkedListNode* head) {
    SinglyLinkedListNode *tortoise, *haire;

    tortoise = head, haire = head;
    while (haire != NULL)
    {
        haire = haire->next;
        if (haire != NULL)
            haire = haire->next, tortoise = tortoise->next;
        if (haire == tortoise)
            return true;
    }
    return false;
}

File: linked_list/hasCycle.py
#!/usr/bin/python3
"""LeetCode Problem #141 --> Linked List Cycle"""


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        tortoise = haire = head

        while haire:
            haire = haire.next

            if haire:
                tortoise = tortoise.next
                haire = haire.next

            if haire is tortoise:
                return True
        return False
